## Phase 2: EDA

#### Reverifying Data

Shape check

Data types

Missing value distribution

Duplicate check

In [38]:
import pandas as pd
import matplotlib 

In [23]:
raw_df = pd.read_csv("itunes_music_cleaned_data.csv")

In [24]:
raw_df.shape

(4913, 9)

In [25]:
raw_df.dtypes

track_id               int64
track_name            object
artist_name           object
collection_name       object
genre                 object
track_price          float64
collection_price     float64
release_date          object
track_time_millis    float64
dtype: object

In [27]:
# Re-apply correct datatypes
raw_df['release_date'] = pd.to_datetime(raw_df['release_date'])
# raw_df['track_time_millis'] = pd.to_numeric(raw_df['track_time_millis'])


# raw_df['track_price'] = pd.to_numeric(raw_df['track_price'])
# raw_df['collection_price'] = pd.to_numeric(raw_df['collection_price'])
# # 


In [28]:
raw_df.dtypes

track_id                           int64
track_name                        object
artist_name                       object
collection_name                   object
genre                             object
track_price                      float64
collection_price                 float64
release_date         datetime64[ns, UTC]
track_time_millis                float64
dtype: object

In [11]:
raw_df.isna().sum()

track_id               0
track_name             0
artist_name            2
collection_name        0
genre                  0
track_price          375
collection_price     376
release_date          49
track_time_millis      1
dtype: int64

In [29]:
raw_df.duplicated().sum()

np.int64(0)

In [30]:
raw_df.head(15)

,track_id,track_name,artist_name,collection_name,genre,track_price,collection_price,release_date,track_time_millis
0,1624932774,POP!,NAYEON,IM NAYEON,Dance,1.29,6.99,2022-06-24 12:00:00+00:00,168107.0
1,1022164261,Riptide,Vance Joy,Dream Your Life Away (Special Edition),Alternative,1.29,9.99,2013-03-22 07:00:00+00:00,202019.0
2,1207120448,Something Just Like This,The Chainsmokers & Coldplay,Memories...Do Not Open,Dance,1.29,9.99,2017-02-22 08:00:00+00:00,247165.0
3,1369380479,lovely,Billie Eilish & Khalid,lovely - Single,Pop,1.29,1.29,2018-04-19 12:00:00+00:00,200187.0
4,1450695872,when the party's over,Billie Eilish,"WHEN WE ALL FALL ASLEEP, WHERE DO WE GO?",Alternative,1.29,10.99,2018-10-17 07:00:00+00:00,196077.0
5,1440899467,ocean eyes,Billie Eilish,dont smile at me,Alternative,1.29,8.99,2016-05-18 12:00:00+00:00,200379.0
6,1468058171,Cruel Summer,Taylor Swift,Lover,Pop,1.29,11.99,2019-08-23 07:00:00+00:00,178426.0
7,303178994,Pop,*NSYNC,Celebrity,Pop,1.29,9.99,2001-07-24 07:00:00+00:00,238427.0
8,1691699836,Lose Control,Teddy Swims,I've Tried Everything But Therapy (Part 1),Pop,1.29,9.99,2023-06-23 12:00:00+00:00,210689.0
9,573962555,When I Was Your Man,Bruno Mars,Unorthodox Jukebox,Pop,1.29,10.99,2012-12-07 08:00:00+00:00,213827.0


In [31]:
raw_df.columns

Index(['track_id', 'track_name', 'artist_name', 'collection_name', 'genre',
       'track_price', 'collection_price', 'release_date', 'track_time_millis'],
      dtype='object')

In [32]:
df = raw_df

## PHASE 2 — EDA BEGINS
##### Numerical Summary

#### Descriptive Statistics

In [33]:
df[['track_price', 'collection_price', 'track_time_millis']].describe()

,track_price,collection_price,track_time_millis
count,4538.000000,4537.000000,4.913000e+03
mean,1.197743,8.621219,2.395843e+05
std,0.238768,5.749726,1.206409e+05
min,-1.000000,-1.000000,3.050300e+04
25%,1.290000,4.990000,1.936030e+05
50%,1.290000,9.990000,2.266810e+05
75%,1.290000,10.990000,2.690320e+05
max,1.290000,119.990000,4.676058e+06


Is pricing standardized?

In [34]:
df['track_price'].dropna().value_counts().sort_index()

track_price
-1.00      34
 0.69      93
 0.99     950
 1.29    3461
Name: count, dtype: int64

their are 34 rows having -1 values
Fix Invalid Prices

In [35]:
df.loc[df['track_price'] == -1, 'track_price'] = None

In [36]:
df['track_price'].value_counts().sort_index()

track_price
0.69      93
0.99     950
1.29    3461
Name: count, dtype: int64

#### Insight 1 — Highly Standardized Pricing

~77% of tracks (3461 / ~4500 non-null) are priced at $1.29

$0.99 is secondary pricing tier

$0.69 is rare (likely promotional or older pricing)

Summary Statistics for Track Price

In [41]:
df['track_price'].describe()

count    4504.000000
mean        1.214334
std         0.143860
min         0.690000
25%         1.290000
50%         1.290000
75%         1.290000
max         1.290000
Name: track_price, dtype: float64

#### Check Unique Collection Prices

In [42]:
df['collection_price'].dropna().value_counts().sort_index()

collection_price
-1.00        26
 0.69        25
 0.99       265
 1.29       569
 1.98         7
 1.99        13
 2.49         1
 2.58         9
 2.97         5
 2.99        18
 3.45         2
 3.49         2
 3.87         6
 3.96        21
 3.99        78
 4.95        37
 4.99       111
 5.16        12
 5.94        18
 5.99       165
 6.15         6
 6.45        17
 6.93        10
 6.99        93
 7.74        33
 7.92        12
 7.99       202
 8.91         6
 8.99       146
 9.03        31
 9.72         5
 9.90        24
 9.99      1267
 10.89        1
 10.99      246
 11.61       11
 11.88        1
 11.99      357
 12.90        3
 12.99      217
 13.86        3
 13.99       85
 14.99       84
 15.99      104
 16.99       79
 17.99        4
 18.99        8
 19.99       32
 20.79        3
 23.76        2
 23.99       16
 24.75        3
 24.99       21
 27.99        3
 29.99        4
 39.99        1
 49.99        3
 69.99        1
 99.99        1
 119.99       2
Name: count, dtype: int

Fixing Invalid Collection Prices

In [43]:
df.loc[df['collection_price'] == -1, 'collection_price'] = None

In [44]:
df['collection_price'].dropna().value_counts().sort_index()

collection_price
0.69        25
0.99       265
1.29       569
1.98         7
1.99        13
2.49         1
2.58         9
2.97         5
2.99        18
3.45         2
3.49         2
3.87         6
3.96        21
3.99        78
4.95        37
4.99       111
5.16        12
5.94        18
5.99       165
6.15         6
6.45        17
6.93        10
6.99        93
7.74        33
7.92        12
7.99       202
8.91         6
8.99       146
9.03        31
9.72         5
9.90        24
9.99      1267
10.89        1
10.99      246
11.61       11
11.88        1
11.99      357
12.90        3
12.99      217
13.86        3
13.99       85
14.99       84
15.99      104
16.99       79
17.99        4
18.99        8
19.99       32
20.79        3
23.76        2
23.99       16
24.75        3
24.99       21
27.99        3
29.99        4
39.99        1
49.99        3
69.99        1
99.99        1
119.99       2
Name: count, dtype: int64

In [45]:
df['collection_price'].describe()

count    4511.000000
mean        8.676673
std         5.719546
min         0.690000
25%         4.990000
50%         9.990000
75%        10.990000
max       119.990000
Name: collection_price, dtype: float64

#### Check Duration Analysis

In [47]:
df['track_time_millis'].describe()

count    4.913000e+03
mean     2.395843e+05
std      1.206409e+05
min      3.050300e+04
25%      1.936030e+05
50%      2.266810e+05
75%      2.690320e+05
max      4.676058e+06
Name: track_time_millis, dtype: float64

Major Observation

Max = 4,676,058 ms (~78 minutes)

That is NOT a normal song.

That is likely:

-Podcast

-Live session

-Full album mistakenly categorized as track

-Data anomaly

Median ≈ 3.8 minutes
Mean ≈ 4 minutes

This tells us:

 Typical track length = 3–4.5 minutes
 Distribution is slightly right-skewed (long tracks pull mean up)

## Detect Duration Outliers 

In [50]:
df[df['track_time_millis'] > 600000][['track_name', 'track_time_millis']].sort_values(by='track_time_millis', ascending=False).head(40)

,track_name,track_time_millis
4447,High Rated Gabru 52 Non Stop Hits,4676058.0
4464,Bala Bala Non Stop Remix,3052014.0
2481,Workout (45 Minutes),2753646.0
4039,Sundar Kand,2750600.0
4662,"Mahamrityunjay Mantra Jaap (From ""Mahamrityunj...",2430698.0
4749,Kaun Gali Gayo Shyam (feat. Na),1409627.0
4712,Apamarjana Stothram (feat. N/A),1159019.0
4751,"Adhyayam, Pt. 1 (feat. N/A)",1095585.0
3775,Maha Mrityunjaya Mantra,938085.0
4708,Devi Kavacham (feat. N/A),749415.0


#### Insight-1  — Dataset Is Not Pure “Songs”

This dataset contains:

-Non-stop remix compilations

-Workout mixes

-Devotional chants

-Long-form audio


#### Insight 2 — Why This Matters

If we calculate:

    -Average duration by genre

    -Median duration

These long tracks will distort mean heavily.

So for fair music analysis, we may need:

    A filtered dataset (tracks < 10 minutes)

### How Many Tracks Are > 10 Minutes?


In [51]:
(df['track_time_millis'] > 600000).sum()

np.int64(14)

Only 14 tracks are longer than 10 minutes.

Out of ~4,913 tracks, that is:
≈0.28%

So these are true outliers, not structural bias.

### Create Clean Dataset having less than 10 min duration song tracks

In [93]:
df_clean = df[df['track_time_millis'] <= 600000]
df_clean['track_time_millis'].describe()

count      4524.000000
mean     235423.862732
std       66936.332315
min       30503.000000
25%      193684.750000
50%      226673.500000
75%      269028.250000
max      588411.000000
Name: track_time_millis, dtype: float64

### After Removing 14 Outliers
1] Mean dropped
    From 3.99 min → 3.92 min
    
2] Standard deviation cut almost in half
    120k → 66k

    That means:
    ✔ Distribution is now tighter
    ✔ Less skew

3] Maximum now realistic

    588,411 ms ≈ 9.8 minutes
    This is acceptable for long songs.

Question:

Do longer songs cost more?

In [94]:
df_clean[['track_price', 'track_time_millis']].corr()

,track_price,track_time_millis
track_price,1.000000,0.020573
track_time_millis,0.020573,1.000000


In [95]:
df_clean[['collection_price', 'track_time_millis']].corr()

,collection_price,track_time_millis
collection_price,1.000000,0.059423
track_time_millis,0.059423,1.000000


In [96]:
# 1. Shape of dataset
print("Shape:", df_clean.shape)

# 2. Missing values
print("\nMissing Values:\n", df_clean.isnull().sum())

# 3. Data types
print("\nData Types:\n", df_clean.dtypes)

# 4. Duplicates
print("\nDuplicate Rows:", df_clean.duplicated().sum())

# 5. Basic numerical summary
print("\nNumerical Summary:\n", df_clean.describe())

# 6. Logical checks
print("\nNegative Track Prices:", (df_clean['track_price'] < 0).sum())
print("Zero Duration Tracks:", (df_clean['track_time_millis'] <= 0).sum())

Shape: (4524, 9)

Missing Values:
 track_id              0
track_name            0
artist_name           2
collection_name       0
genre                 0
track_price          20
collection_price     27
release_date         46
track_time_millis     0
dtype: int64

Data Types:
 track_id                           int64
track_name                        object
artist_name                       object
collection_name                   object
genre                             object
track_price                      float64
collection_price                 float64
release_date         datetime64[ns, UTC]
track_time_millis                float64
dtype: object

Duplicate Rows: 0

Numerical Summary:
            track_id  track_price  collection_price  track_time_millis
count  4.524000e+03  4504.000000       4497.000000        4524.000000
mean   1.341292e+09     1.214334          8.680145      235423.862732
std    4.313947e+08     0.143860          5.726679       66936.332315
min    6.920394e+06

#### Drop rows where price is missing

In [97]:
df_clean = df_clean.dropna(subset=['track_price', 'collection_price'])
df_clean = df_clean.dropna() ## Drop all rows containing any missing value



#### track_time_millis is still float64.

Duration should ideally be integer.

In [98]:
df_clean['track_time_millis'] = df_clean['track_time_millis'].astype(int)

In [99]:
print("Data Type of track time :", df_clean["track_time_millis"].dtypes)

Data Type of track time : int64


# Complete Data Re-Verification

In [110]:
# FINAL DATA RE-VERIFICATION

print("------ DATASET SHAPE ------")
print("Shape:", df_clean.shape)


print("\n------ MISSING VALUES ------")
print(df_clean.isnull().sum())


print("\n------ DATA TYPES ------")
print(df_clean.dtypes)


print("\n------ DUPLICATE CHECK ------")
print("Duplicate Rows:", df_clean.duplicated().sum())


print("\n------ NUMERICAL SUMMARY ------")
print(df_clean.describe())


print("\n------ LOGICAL VALIDATION ------")
print("Negative Track Prices:", (df_clean['track_price'] < 0).sum())
print("Negative Collection Prices:", (df_clean['collection_price'] < 0).sum())
print("Zero or Negative Duration:", (df_clean['track_time_millis'] <= 0).sum())


print("\n------ PRICE TIER CHECK ------")
print("Unique Track Prices:", sorted(df_clean['track_price'].unique()))

------ DATASET SHAPE ------
Shape: (4429, 9)

------ MISSING VALUES ------
track_id             0
track_name           0
artist_name          0
collection_name      0
genre                0
track_price          0
collection_price     0
release_date         0
track_time_millis    0
dtype: int64

------ DATA TYPES ------
track_id                           int64
track_name                        object
artist_name                       object
collection_name                   object
genre                             object
track_price                      float64
collection_price                 float64
release_date         datetime64[ns, UTC]
track_time_millis                  int64
dtype: object

------ DUPLICATE CHECK ------
Duplicate Rows: 0

------ NUMERICAL SUMMARY ------
           track_id  track_price  collection_price  track_time_millis
count  4.429000e+03  4429.000000       4429.000000        4429.000000
mean   1.338347e+09     1.216236          8.633843      235908.071574
std 

In [102]:
total_rows = len(df_clean)

track_price_missing_pct = (df_clean['track_price'].isna().sum() / total_rows) * 100
collection_price_missing_pct = (df_clean['collection_price'].isna().sum() / total_rows) * 100

print(f"Track Price Missing %: {track_price_missing_pct:.2f}%")
print(f"Collection Price Missing %: {collection_price_missing_pct:.2f}%")

Track Price Missing %: 0.00%
Collection Price Missing %: 0.00%


removing rows where both track_price AND collection_price are missing.

In [109]:
print(df_clean.isnull().sum())


track_id             0
track_name           0
artist_name          0
collection_name      0
genre                0
track_price          0
collection_price     0
release_date         0
track_time_millis    0
dtype: int64


Now Our data have 

    -0 missing values

    -0 duplicates

    -No negative pricing

    -No zero/negative duration

## checking: Album Price Consistency & Album Size Distribution

Does the same album (collection_name) appear with multiple different collection_price values?

In [104]:
album_validation = (
    df.groupby('collection_name')
      .agg(
          tracks_per_album=('track_id', 'count'),
          unique_collection_prices=('collection_price', 'nunique'),
          collection_price=('collection_price', 'first')
      )
      .reset_index()
)

album_validation.head()

,collection_name,tracks_per_album,unique_collection_prices,collection_price
0,#1's: Destiny's Child,1,1,9.99
1,#NAME?,13,3,11.99
2,#willpower (Deluxe),1,1,11.99
3,$ome $exy $ongs 4 U,3,1,11.99
4,&ME Edition,1,1,15.99


In [107]:
print("Albums with multiple collection prices:",
      (album_validation['unique_collection_prices'] > 1).sum())

print("\nMean tracks per album:",
      album_validation['tracks_per_album'].mean())

print("Albums with 1 track:",
      (album_validation['tracks_per_album'] == 1).sum())

print("Albums with 100+ tracks:",
      (album_validation['tracks_per_album'] >= 100).sum())

max_price = album_validation['collection_price'].max()

print("\nMax Collection Price:", max_price)

print("Track count for max price album:")
print(
    album_validation[
        album_validation['collection_price'] == max_price
    ][['tracks_per_album', 'collection_price']]
)

Albums with multiple collection prices: 24

Mean tracks per album: 1.6857355126300149
Albums with 1 track: 2133
Albums with 100+ tracks: 0

Max Collection Price: 119.99
Track count for max price album:
      tracks_per_album  collection_price
990                  1            119.99
1465                 1            119.99


In [111]:
df_clean.to_csv("itunes_EDA_cleaned_analytics_3.csv", index=False)